// echarts-for-angular/esm2020/lib/driective/EchartsDirective.mjs
import { Directive, EventEmitter, Input, Output } from "@angular/core";
import * as echarts from 'echarts/core';
import { CanvasRenderer } from 'echarts/renderers';
import { HtmlHelper } from "../internal/HtmlHelper";
import * as i0 from "@angular/core";
export class EchartsDirective {
    constructor(_el) {
        this._el = _el;
        this.extentions = [];
        this.isResizable = true;
        this.defaultWidth = 400;
        this.defaultHeight = 400;
        this.periodicityInMiliSeconds = 2000;
        this.theme = '';
        this.chartInit = new EventEmitter();
    }
    ngOnInit() {
        echarts.use([...this.extentions, CanvasRenderer]);
        this._echartsInstance = echarts.init(this._el.nativeElement, this.theme, {
            width: this._el.nativeElement.clientWidth === this.defaultWidth ? 400 : undefined,
            height: this._el.nativeElement.clientHeight === 0 ? this.defaultHeight : undefined
        });
        this.chartInit.emit(this._echartsInstance);
        this._setParams();
        if (this.isResizable) {
            this._addResizbleFunctionality();
        }
    }
    ngOnChanges(changes) {
        if (changes.options && !changes.options.firstChange) {
            this._setParams();
        }
        if (changes.isResizable && !changes.isResizable.firstChange) {
            if (this.isResizable) {
                this._addResizbleFunctionality();
            }
            else {
                if (this._subscription != null)
                    this._subscription.unsubscribe();
            }
        }
    }
    _addResizbleFunctionality() {
        if (this._subscription != null)
            this._subscription.unsubscribe();
        this._subscription = HtmlHelper.getWidthSensor(this._el.nativeElement).subscribe(() => {
            if (this._echartsInstance != null) {
                this._echartsInstance.resize();
            }
        });
    }
    _setParams() {
        if (this._echartsInstance != null && this.options != null) {
            this._echartsInstance.setOption(this.options, true);
        }
    }
    ngOnDestroy() {
        if (this._subscription != null)
            this._subscription.unsubscribe();
    }
}
EchartsDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.8", ngImport: i0, type: EchartsDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
EchartsDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.8", type: EchartsDirective, selector: "[echarts]", inputs: { options: "options", extentions: "extentions", isResizable: "isResizable", defaultWidth: "defaultWidth", defaultHeight: "defaultHeight", periodicityInMiliSeconds: "periodicityInMiliSeconds", theme: "theme" }, outputs: { chartInit: "chartInit" }, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.8", ngImport: i0, type: EchartsDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: "[echarts]"
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; }, propDecorators: { options: [{
                type: Input
            }], extentions: [{
                type: Input
            }], isResizable: [{
                type: Input
            }], defaultWidth: [{
                type: Input
            }], defaultHeight: [{
                type: Input
            }], periodicityInMiliSeconds: [{
                type: Input
            }], theme: [{
                type: Input
            }], chartInit: [{
                type: Output
            }] } });
