// angular-calendar/esm2015/modules/common/keydown-enter.directive.js
import { Directive, Output, EventEmitter, ElementRef, NgZone, Renderer2, } from '@angular/core';
export class KeydownEnterDirective {
    constructor(host, ngZone, renderer) {
        this.host = host;
        this.ngZone = ngZone;
        this.renderer = renderer;
        this.keydown = new EventEmitter(); // eslint-disable-line
        this.keydownListener = null;
    }
    ngOnInit() {
        this.ngZone.runOutsideAngular(() => {
            this.keydownListener = this.renderer.listen(this.host.nativeElement, 'keydown', (event) => {
                if (event.keyCode === 13 ||
                    event.which === 13 ||
                    event.key === 'Enter') {
                    event.preventDefault();
                    event.stopPropagation();
                    this.ngZone.run(() => {
                        this.keydown.emit(event);
                    });
                }
            });
        });
    }
    ngOnDestroy() {
        if (this.keydownListener !== null) {
            this.keydownListener();
            this.keydownListener = null;
        }
    }
}
KeydownEnterDirective.decorators = [
    { type: Directive, args: [{
                selector: '[mwlKeydownEnter]',
            },] }
];
KeydownEnterDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: NgZone },
    { type: Renderer2 }
];
KeydownEnterDirective.propDecorators = {
    keydown: [{ type: Output, args: ['mwlKeydownEnter',] }]
};
